File: mobile/android/app/src/main/java/com/example/seniordesignmobileapp/domain/PostureAnalysisCoordinator.kt
package com.example.seniordesignmobileapp.domain

import com.example.seniordesignmobileapp.analysis.AnalysisConfig
import com.example.seniordesignmobileapp.analysis.AnalysisInputWindow
import com.example.seniordesignmobileapp.analysis.AnalysisWindowSummary
import com.example.seniordesignmobileapp.analysis.PostureAlert
import com.example.seniordesignmobileapp.analysis.PostureAlertCode
import com.example.seniordesignmobileapp.analysis.PostureAnalysisResult
import com.example.seniordesignmobileapp.analysis.PostureAnalyzer
import com.example.seniordesignmobileapp.analysis.SensorAssignment
import com.example.seniordesignmobileapp.analysis.SensorPlacement
import com.example.seniordesignmobileapp.analysis.SittingCalibration
import com.example.seniordesignmobileapp.analysis.SittingPostureAnalyzer
import com.example.seniordesignmobileapp.analysis.SittingPostureMath
import com.example.seniordesignmobileapp.analysis.WindowSpec
import com.example.seniordesignmobileapp.model.ImuSample
import com.example.seniordesignmobileapp.model.NetworkStatus

class PostureAnalysisCoordinator(
    private val config: AnalysisConfig = AnalysisConfig.sittingDefault(),
    private val analyzer: PostureAnalyzer = SittingPostureAnalyzer(),
) {
    private val bufferedSamplesBySensor = mutableMapOf<Int, ArrayDeque<BufferedSample>>()
    private val observedSensorIds = linkedSetOf<Int>()
    private var latestNetworkStatus: NetworkStatus? = null
    private var sittingCalibration: SittingCalibration? = null
    private var calibrationMessage: String? = null
    private var manualUpperBackSensorId: Int? = null
    private var manualLowerBackSensorId: Int? = null

    fun onSample(
        receivedAtEpochMs: Long,
        sample: ImuSample,
    ): PostureAnalysisSnapshot {
        observedSensorIds += sample.sensorId
        bufferedSamplesBySensor.getOrPut(sample.sensorId) { ArrayDeque() }
            .addLast(BufferedSample(receivedAtEpochMs, sample))
        return buildSnapshot(nowEpochMs = receivedAtEpochMs)
    }

    fun onNetworkStatus(
        receivedAtEpochMs: Long,
        status: NetworkStatus,
    ): PostureAnalysisSnapshot {
        latestNetworkStatus = status
        observedSensorIds += status.sensors.map { it.sensorId }
        return buildSnapshot(nowEpochMs = receivedAtEpochMs)
    }

    fun captureCalibration(
        nowEpochMs: Long,
    ): CalibrationCaptureResult {
        pruneExpiredSamples(nowEpochMs)
        val sensorAssignments = resolveSensorAssignments()
        val upperBackSensorId = sensorAssignments
            .firstOrNull { it.placement == SensorPlacement.UPPER_BACK }
            ?.sensorId
        val lowerBackSensorId = sensorAssignments
            .firstOrNull { it.placement == SensorPlacement.LOWER_BACK }
            ?.sensorId
        val upperBackMetrics = upperBackSensorId?.let { sensorId ->
            SittingPostureMath.computeSensorMetrics(bufferedSamplesBySensor[sensorId].orEmpty().map { it.sample })
        }
        val lowerBackMetrics = lowerBackSensorId?.let { sensorId ->
            SittingPostureMath.computeSensorMetrics(bufferedSamplesBySensor[sensorId].orEmpty().map { it.sample })
        }

        if (upperBackSensorId == null || lowerBackSensorId == null || upperBackMetrics == null || lowerBackMetrics == null) {
            calibrationMessage = "Calibration needs both upper and lower back sensors with live samples."
            return CalibrationCaptureResult(
                success = false,
                message = calibrationMessage!!,
                snapshot = buildSnapshot(nowEpochMs),
            )
        }

        sittingCalibration = SittingCalibration(
            capturedAtEpochMs = nowEpochMs,
            upperBackSensorId = upperBackSensorId,
            lowerBackSensorId = lowerBackSensorId,
            upperBackPitchDeg = upperBackMetrics.pitchDeg,
            lowerBackPitchDeg = lowerBackMetrics.pitchDeg,
            bendAngleDeg = SittingPostureMath.angleBetween(
                upperBackMetrics.gravityVector,
                lowerBackMetrics.gravityVector,
            ),
        )
        calibrationMessage = "Captured upright sitting calibration using sensors $upperBackSensorId and $lowerBackSensorId."
        return CalibrationCaptureResult(
            success = true,
            message = calibrationMessage!!,
            snapshot = buildSnapshot(nowEpochMs),
        )
    }

    fun setUpperBackSensor(
        sensorId: Int?,
        nowEpochMs: Long,
    ): PostureAnalysisSnapshot {
        if (manualUpperBackSensorId != sensorId) {
            manualUpperBackSensorId = sensorId
            clearCalibration("Upper back sensor selection updated. Recalibrate sitting posture.")
        }
        return buildSnapshot(nowEpochMs)
    }

    fun setLowerBackSensor(
        sensorId: Int?,
        nowEpochMs: Long,
    ): PostureAnalysisSnapshot {
        if (manualLowerBackSensorId != sensorId) {
            manualLowerBackSensorId = sensorId
            clearCalibration("Lower back sensor selection updated. Recalibrate sitting posture.")
        }
        return buildSnapshot(nowEpochMs)
    }

    fun clear() {
        bufferedSamplesBySensor.clear()
        observedSensorIds.clear()
        latestNetworkStatus = null
        sittingCalibration = null
        calibrationMessage = null
        manualUpperBackSensorId = null
        manualLowerBackSensorId = null
    }

    fun currentSnapshot(
        nowEpochMs: Long,
    ): PostureAnalysisSnapshot = buildSnapshot(nowEpochMs)

    private fun buildSnapshot(
        nowEpochMs: Long,
    ): PostureAnalysisSnapshot {
        pruneExpiredSamples(nowEpochMs)

        val sensorAssignments = resolveSensorAssignments()
        val effectiveConfig = config.copy(expectedSensors = sensorAssignments)
        val windowDurationMs = effectiveConfig.windowSpec.maxDurationMs()
        val lookbackMs = effectiveConfig.historyLookbackMs
        val windowEndEpochMs = nowEpochMs - lookbackMs
        val windowStartEpochMs = windowEndEpochMs - windowDurationMs
        val minWindowDurationMs = effectiveConfig.windowSpec.minDurationMs()

        val timedSamplesBySensor = bufferedSamplesBySensor.mapValues { (_, samples) ->
            samples.filter { it.receivedAtEpochMs in windowStartEpochMs..windowEndEpochMs }
        }.filterValues { it.isNotEmpty() }
        val samplesBySensor = timedSamplesBySensor.mapValues { (_, samples) ->
            samples.map { it.sample }
        }

        val expectedSensors = sensorAssignments
            .filter { it.required }
            .map { it.sensorId }
            .toSet()
        val availableSensors = samplesBySensor.keys
        val missingSensors = expectedSensors - availableSensors

        val windowSummary = AnalysisWindowSummary(
            windowStartEpochMs = windowStartEpochMs,
            windowEndEpochMs = windowEndEpochMs,
            lookbackMs = lookbackMs,
            sampleCountsBySensor = samplesBySensor.mapValues { it.value.size },
            availableSensors = availableSensors,
            missingSensors = missingSensors,
        )

        if (samplesBySensor.isEmpty()) {
            return PostureAnalysisSnapshot(
                config = effectiveConfig,
                sensorAssignments = sensorAssignments,
                availableSensorIds = resolveAvailableSensorIds(),
                manualUpperBackSensorId = manualUpperBackSensorId,
                manualLowerBackSensorId = manualLowerBackSensorId,
                expectedSensors = expectedSensors,
                expectedSensorsInferred = config.expectedSensors.isEmpty(),
                sittingCalibration = sittingCalibration,
                windowSummary = windowSummary,
                latestResult = null,
                calibrationMessage = calibrationMessage,
                statusMessage = "Waiting for enough live samples to build an analysis window.",
            )
        }

        val inputWindow = AnalysisInputWindow(
            windowStartEpochMs = windowStartEpochMs,
            windowEndEpochMs = windowEndEpochMs,
            lookbackMs = lookbackMs,
            samplesBySensor = samplesBySensor,
            networkStatus = latestNetworkStatus,
            sensorAssignments = sensorAssignments,
            expectedSensors = expectedSensors,
            availableSensors = availableSensors,
            missingSensors = missingSensors,
            sittingCalibration = sittingCalibration,
        )

        val baseResult = analyzer.analyze(inputWindow, effectiveConfig)
        val coverageMs = computeCoverageMs(
            timedSamplesBySensor = timedSamplesBySensor,
            windowStartEpochMs = windowStartEpochMs,
            windowEndEpochMs = windowEndEpochMs,
        )
        val augmentedAlerts = buildList {
            addAll(baseResult.alerts)
            if (coverageMs < minWindowDurationMs) {
                add(
                    PostureAlert(
                        code = PostureAlertCode.INSUFFICIENT_WINDOW,
                        message = "Only ${coverageMs} ms of history is buffered; ${minWindowDurationMs} ms minimum is preferred.",
                    )
                )
            }
        }.distinct()
        val result = baseResult.copy(alerts = augmentedAlerts)
        val statusMessage = when {
            sensorAssignments.size < 2 -> "Waiting to identify two back sensors for sitting analysis."
            coverageMs < minWindowDurationMs -> "Collecting more sample history for the requested analysis window."
            sittingCalibration == null -> "Capture an upright sitting calibration to start scoring."
            missingSensors.isNotEmpty() -> "Partial posture preview only; both back sensors are needed for a score."
            else -> "Calibrated sitting analysis is active."
        }

        return PostureAnalysisSnapshot(
            config = effectiveConfig,
            sensorAssignments = sensorAssignments,
            availableSensorIds = resolveAvailableSensorIds(),
            manualUpperBackSensorId = manualUpperBackSensorId,
            manualLowerBackSensorId = manualLowerBackSensorId,
            expectedSensors = expectedSensors,
            expectedSensorsInferred = config.expectedSensors.isEmpty(),
            sittingCalibration = sittingCalibration,
            windowSummary = windowSummary,
            latestResult = result,
            calibrationMessage = calibrationMessage,
            statusMessage = statusMessage,
        )
    }

    private fun resolveSensorAssignments(): List<SensorAssignment> {
        if (config.expectedSensors.isNotEmpty()) {
            return config.expectedSensors
        }

        val availableSensorIds = resolveAvailableSensorIds().toMutableList()
        val assignments = mutableListOf<SensorAssignment>()

        val lowerBackSensorId = resolveSensorForPlacement(
            preferredSensorId = manualLowerBackSensorId,
            availableSensorIds = availableSensorIds,
        )
        if (lowerBackSensorId != null) {
            assignments += SensorAssignment(
                sensorId = lowerBackSensorId,
                placement = SensorPlacement.LOWER_BACK,
            )
            availableSensorIds.remove(lowerBackSensorId)
        }

        val upperBackSensorId = resolveSensorForPlacement(
            preferredSensorId = manualUpperBackSensorId,
            availableSensorIds = availableSensorIds,
        )
        if (upperBackSensorId != null) {
            assignments += SensorAssignment(
                sensorId = upperBackSensorId,
                placement = SensorPlacement.UPPER_BACK,
            )
        }

        return assignments
    }

    private fun resolveAvailableSensorIds(): List<Int> {
        val liveSensorIds = when {
            latestNetworkStatus?.sensors?.isNotEmpty() == true -> latestNetworkStatus!!.sensors
                .map { it.sensorId }
                .sorted()

            observedSensorIds.isNotEmpty() -> observedSensorIds.sorted()
            else -> emptyList()
        }

        return (liveSensorIds + listOfNotNull(manualLowerBackSensorId, manualUpperBackSensorId))
            .distinct()
            .sorted()
    }

    private fun resolveSensorForPlacement(
        preferredSensorId: Int?,
        availableSensorIds: List<Int>,
    ): Int? {
        if (preferredSensorId != null && preferredSensorId in availableSensorIds) {
            return preferredSensorId
        }
        return availableSensorIds.firstOrNull()
    }

    private fun clearCalibration(
        message: String,
    ) {
        sittingCalibration = null
        calibrationMessage = message
    }

    private fun pruneExpiredSamples(nowEpochMs: Long) {
        val retentionStartEpochMs = nowEpochMs - config.windowSpec.maxDurationMs() - config.historyLookbackMs
        val iterator = bufferedSamplesBySensor.iterator()
        while (iterator.hasNext()) {
            val (_, samples) = iterator.next()
            while (samples.isNotEmpty() && samples.first().receivedAtEpochMs < retentionStartEpochMs) {
                samples.removeFirst()
            }
            if (samples.isEmpty()) {
                iterator.remove()
            }
        }
    }

    private fun computeCoverageMs(
        timedSamplesBySensor: Map<Int, List<BufferedSample>>,
        windowStartEpochMs: Long,
        windowEndEpochMs: Long,
    ): Long {
        val sampleTimes = timedSamplesBySensor.values.flatten().map { it.receivedAtEpochMs }
        if (sampleTimes.isEmpty()) {
            return 0L
        }
        val earliestSampleEpochMs = sampleTimes.minOrNull() ?: return 0L
        return (windowEndEpochMs - maxOf(windowStartEpochMs, earliestSampleEpochMs)).coerceAtLeast(0L)
    }
}

data class CalibrationCaptureResult(
    val success: Boolean,
    val message: String,
    val snapshot: PostureAnalysisSnapshot,
)

data class PostureAnalysisSnapshot(
    val config: AnalysisConfig,
    val sensorAssignments: List<SensorAssignment>,
    val availableSensorIds: List<Int>,
    val manualUpperBackSensorId: Int?,
    val manualLowerBackSensorId: Int?,
    val expectedSensors: Set<Int>,
    val expectedSensorsInferred: Boolean,
    val sittingCalibration: SittingCalibration?,
    val windowSummary: AnalysisWindowSummary,
    val latestResult: PostureAnalysisResult?,
    val calibrationMessage: String?,
    val statusMessage: String,
)

private data class BufferedSample(
    val receivedAtEpochMs: Long,
    val sample: ImuSample,
)

private fun WindowSpec.maxDurationMs(): Long =
    when (this) {
        is WindowSpec.FixedDuration -> durationMs
        is WindowSpec.DurationRange -> maxDurationMs
    }

private fun WindowSpec.minDurationMs(): Long =
    when (this) {
        is WindowSpec.FixedDuration -> durationMs
        is WindowSpec.DurationRange -> minDurationMs
    }

File: mobile/ios/FH07Mobile/Domain/PostureAnalysisCoordinator.swift
import Foundation

struct CalibrationCaptureResult {
    let success: Bool
    let message: String
    let snapshot: PostureAnalysisSnapshot
}

struct PostureAnalysisSnapshot {
    let config: AnalysisConfig
    let sensorAssignments: [SensorAssignment]
    let availableSensorIds: [Int]
    let manualUpperBackSensorId: Int?
    let manualLowerBackSensorId: Int?
    let expectedSensors: Set<Int>
    let expectedSensorsInferred: Bool
    let sittingCalibration: SittingCalibration?
    let windowSummary: AnalysisWindowSummary
    let latestResult: PostureAnalysisResult?
    let calibrationMessage: String?
    let statusMessage: String
}

final class PostureAnalysisCoordinator {
    private let config: AnalysisConfig
    private let analyzer: PostureAnalyzer

    private var bufferedSamplesBySensor: [Int: [BufferedSample]] = [:]
    private var observedSensorIds: [Int] = [] // insertion-ordered unique set
    private var latestNetworkStatus: NetworkStatus? = nil
    private var sittingCalibration: SittingCalibration? = nil
    private var calibrationMessage: String? = nil
    private var manualUpperBackSensorId: Int? = nil
    private var manualLowerBackSensorId: Int? = nil

    init(
        config: AnalysisConfig = AnalysisConfig.sittingDefault(),
        analyzer: PostureAnalyzer = SittingPostureAnalyzer()
    ) {
        self.config = config
        self.analyzer = analyzer
    }

    func onSample(receivedAtEpochMs: Int64, sample: ImuSample) -> PostureAnalysisSnapshot {
        if !observedSensorIds.contains(sample.sensorId) {
            observedSensorIds.append(sample.sensorId)
        }
        bufferedSamplesBySensor[sample.sensorId, default: []]
            .append(BufferedSample(receivedAtEpochMs: receivedAtEpochMs, sample: sample))
        return buildSnapshot(nowEpochMs: receivedAtEpochMs)
    }

    func onNetworkStatus(receivedAtEpochMs: Int64, status: NetworkStatus) -> PostureAnalysisSnapshot {
        latestNetworkStatus = status
        for s in status.sensors {
            if !observedSensorIds.contains(s.sensorId) {
                observedSensorIds.append(s.sensorId)
            }
        }
        return buildSnapshot(nowEpochMs: receivedAtEpochMs)
    }

    func captureCalibration(nowEpochMs: Int64) -> CalibrationCaptureResult {
        pruneExpiredSamples(nowEpochMs: nowEpochMs)
        let sensorAssignments = resolveSensorAssignments()
        let upperBackSensorId = sensorAssignments.first(where: { $0.placement == .upperBack })?.sensorId
        let lowerBackSensorId = sensorAssignments.first(where: { $0.placement == .lowerBack })?.sensorId
        let upperBackMetrics: SensorWindowMetrics? = upperBackSensorId.flatMap { id in
            SittingPostureMath.computeSensorMetrics(samples: (bufferedSamplesBySensor[id] ?? []).map { $0.sample })
        }
        let lowerBackMetrics: SensorWindowMetrics? = lowerBackSensorId.flatMap { id in
            SittingPostureMath.computeSensorMetrics(samples: (bufferedSamplesBySensor[id] ?? []).map { $0.sample })
        }

        if upperBackSensorId == nil || lowerBackSensorId == nil || upperBackMetrics == nil || lowerBackMetrics == nil {
            calibrationMessage = "Calibration needs both upper and lower back sensors with live samples."
            return CalibrationCaptureResult(
                success: false,
                message: calibrationMessage!,
                snapshot: buildSnapshot(nowEpochMs: nowEpochMs)
            )
        }

        sittingCalibration = SittingCalibration(
            capturedAtEpochMs: nowEpochMs,
            upperBackSensorId: upperBackSensorId!,
            lowerBackSensorId: lowerBackSensorId!,
            upperBackPitchDeg: upperBackMetrics!.pitchDeg,
            lowerBackPitchDeg: lowerBackMetrics!.pitchDeg,
            bendAngleDeg: SittingPostureMath.angleBetween(
                first: upperBackMetrics!.gravityVector,
                second: lowerBackMetrics!.gravityVector
            )
        )
        calibrationMessage = "Captured upright sitting calibration using sensors \(upperBackSensorId!) and \(lowerBackSensorId!)."
        return CalibrationCaptureResult(
            success: true,
            message: calibrationMessage!,
            snapshot: buildSnapshot(nowEpochMs: nowEpochMs)
        )
    }

    func setUpperBackSensor(sensorId: Int?, nowEpochMs: Int64) -> PostureAnalysisSnapshot {
        if manualUpperBackSensorId != sensorId {
            manualUpperBackSensorId = sensorId
            clearCalibration(message: "Upper back sensor selection updated. Recalibrate sitting posture.")
        }
        return buildSnapshot(nowEpochMs: nowEpochMs)
    }

    func setLowerBackSensor(sensorId: Int?, nowEpochMs: Int64) -> PostureAnalysisSnapshot {
        if manualLowerBackSensorId != sensorId {
            manualLowerBackSensorId = sensorId
            clearCalibration(message: "Lower back sensor selection updated. Recalibrate sitting posture.")
        }
        return buildSnapshot(nowEpochMs: nowEpochMs)
    }

    func clear() {
        bufferedSamplesBySensor.removeAll()
        observedSensorIds.removeAll()
        latestNetworkStatus = nil
        sittingCalibration = nil
        calibrationMessage = nil
        manualUpperBackSensorId = nil
        manualLowerBackSensorId = nil
    }

    func currentSnapshot(nowEpochMs: Int64) -> PostureAnalysisSnapshot {
        buildSnapshot(nowEpochMs: nowEpochMs)
    }

    private func buildSnapshot(nowEpochMs: Int64) -> PostureAnalysisSnapshot {
        pruneExpiredSamples(nowEpochMs: nowEpochMs)

        let sensorAssignments = resolveSensorAssignments()
        var effectiveConfig = config
        effectiveConfig.expectedSensors = sensorAssignments
        let windowDurationMs = effectiveConfig.windowSpec.maxDurationMs
        let lookbackMs = effectiveConfig.historyLookbackMs
        let windowEndEpochMs = nowEpochMs - lookbackMs
        let windowStartEpochMs = windowEndEpochMs - windowDurationMs
        let minWindowDurationMs = effectiveConfig.windowSpec.minDurationMs

        var timedSamplesBySensor: [Int: [BufferedSample]] = [:]
        for (sensorId, samples) in bufferedSamplesBySensor {
            let filtered = samples.filter { $0.receivedAtEpochMs >= windowStartEpochMs && $0.receivedAtEpochMs <= windowEndEpochMs }
            if !filtered.isEmpty {
                timedSamplesBySensor[sensorId] = filtered
            }
        }
        let samplesBySensor: [Int: [ImuSample]] = timedSamplesBySensor.mapValues { $0.map { $0.sample } }

        let expectedSensors = Set(sensorAssignments.filter { $0.required }.map { $0.sensorId })
        let availableSensors = Set(samplesBySensor.keys)
        let missingSensors = expectedSensors.subtracting(availableSensors)

        let windowSummary = AnalysisWindowSummary(
            windowStartEpochMs: windowStartEpochMs,
            windowEndEpochMs: windowEndEpochMs,
            lookbackMs: lookbackMs,
            sampleCountsBySensor: samplesBySensor.mapValues { $0.count },
            availableSensors: availableSensors,
            missingSensors: missingSensors
        )

        if samplesBySensor.isEmpty {
            return PostureAnalysisSnapshot(
                config: effectiveConfig,
                sensorAssignments: sensorAssignments,
                availableSensorIds: resolveAvailableSensorIds(),
                manualUpperBackSensorId: manualUpperBackSensorId,
                manualLowerBackSensorId: manualLowerBackSensorId,
                expectedSensors: expectedSensors,
                expectedSensorsInferred: config.expectedSensors.isEmpty,
                sittingCalibration: sittingCalibration,
                windowSummary: windowSummary,
                latestResult: nil,
                calibrationMessage: calibrationMessage,
                statusMessage: "Waiting for enough live samples to build an analysis window."
            )
        }

        let inputWindow = AnalysisInputWindow(
            windowStartEpochMs: windowStartEpochMs,
            windowEndEpochMs: windowEndEpochMs,
            lookbackMs: lookbackMs,
            samplesBySensor: samplesBySensor,
            networkStatus: latestNetworkStatus,
            sensorAssignments: sensorAssignments,
            expectedSensors: expectedSensors,
            availableSensors: availableSensors,
            missingSensors: missingSensors,
            sittingCalibration: sittingCalibration
        )

        var baseResult = analyzer.analyze(input: inputWindow, config: effectiveConfig)
        let coverageMs = computeCoverageMs(
            timedSamplesBySensor: timedSamplesBySensor,
            windowStartEpochMs: windowStartEpochMs,
            windowEndEpochMs: windowEndEpochMs
        )
        var augmentedAlerts = baseResult.alerts
        if coverageMs < minWindowDurationMs {
            let alert = PostureAlert(
                code: .insufficientWindow,
                message: "Only \(coverageMs) ms of history is buffered; \(minWindowDurationMs) ms minimum is preferred."
            )
            if !augmentedAlerts.contains(alert) {
                augmentedAlerts.append(alert)
            }
        }
        baseResult.alerts = augmentedAlerts
        let statusMessage: String
        if sensorAssignments.count < 2 {
            statusMessage = "Waiting to identify two back sensors for sitting analysis."
        } else if coverageMs < minWindowDurationMs {
            statusMessage = "Collecting more sample history for the requested analysis window."
        } else if sittingCalibration == nil {
            statusMessage = "Capture an upright sitting calibration to start scoring."
        } else if !missingSensors.isEmpty {
            statusMessage = "Partial posture preview only; both back sensors are needed for a score."
        } else {
            statusMessage = "Calibrated sitting analysis is active."
        }

        return PostureAnalysisSnapshot(
            config: effectiveConfig,
            sensorAssignments: sensorAssignments,
            availableSensorIds: resolveAvailableSensorIds(),
            manualUpperBackSensorId: manualUpperBackSensorId,
            manualLowerBackSensorId: manualLowerBackSensorId,
            expectedSensors: expectedSensors,
            expectedSensorsInferred: config.expectedSensors.isEmpty,
            sittingCalibration: sittingCalibration,
            windowSummary: windowSummary,
            latestResult: baseResult,
            calibrationMessage: calibrationMessage,
            statusMessage: statusMessage
        )
    }

    private func resolveSensorAssignments() -> [SensorAssignment] {
        if !config.expectedSensors.isEmpty {
            return config.expectedSensors
        }

        var availableSensorIds = resolveAvailableSensorIds()
        var assignments: [SensorAssignment] = []

        if let lowerBackSensorId = resolveSensorForPlacement(
            preferredSensorId: manualLowerBackSensorId,
            availableSensorIds: availableSensorIds
        ) {
            assignments.append(SensorAssignment(sensorId: lowerBackSensorId, placement: .lowerBack))
            availableSensorIds.removeAll(where: { $0 == lowerBackSensorId })
        }

        if let upperBackSensorId = resolveSensorForPlacement(
            preferredSensorId: manualUpperBackSensorId,
            availableSensorIds: availableSensorIds
        ) {
            assignments.append(SensorAssignment(sensorId: upperBackSensorId, placement: .upperBack))
        }

        return assignments
    }

    private func resolveAvailableSensorIds() -> [Int] {
        let liveSensorIds: [Int]
        if let status = latestNetworkStatus, !status.sensors.isEmpty {
            liveSensorIds = status.sensors.map { $0.sensorId }.sorted()
        } else if !observedSensorIds.isEmpty {
            liveSensorIds = observedSensorIds.sorted()
        } else {
            liveSensorIds = []
        }

        var result = liveSensorIds
        if let id = manualLowerBackSensorId { result.append(id) }
        if let id = manualUpperBackSensorId { result.append(id) }
        return Array(Set(result)).sorted()
    }

    private func resolveSensorForPlacement(preferredSensorId: Int?, availableSensorIds: [Int]) -> Int? {
        if let preferred = preferredSensorId, availableSensorIds.contains(preferred) {
            return preferred
        }
        return availableSensorIds.first
    }

    private func clearCalibration(message: String) {
        sittingCalibration = nil
        calibrationMessage = message
    }

    private func pruneExpiredSamples(nowEpochMs: Int64) {
        let retentionStartEpochMs = nowEpochMs - config.windowSpec.maxDurationMs - config.historyLookbackMs
        for (sensorId, samples) in bufferedSamplesBySensor {
            var mutated = samples
            while let first = mutated.first, first.receivedAtEpochMs < retentionStartEpochMs {
                mutated.removeFirst()
            }
            if mutated.isEmpty {
                bufferedSamplesBySensor.removeValue(forKey: sensorId)
            } else {
                bufferedSamplesBySensor[sensorId] = mutated
            }
        }
    }

    private func computeCoverageMs(
        timedSamplesBySensor: [Int: [BufferedSample]],
        windowStartEpochMs: Int64,
        windowEndEpochMs: Int64
    ) -> Int64 {
        let sampleTimes = timedSamplesBySensor.values.flatMap { $0 }.map { $0.receivedAtEpochMs }
        if sampleTimes.isEmpty { return 0 }
        guard let earliestSampleEpochMs = sampleTimes.min() else { return 0 }
        return max(windowEndEpochMs - max(windowStartEpochMs, earliestSampleEpochMs), 0)
    }
}

private struct BufferedSample {
    let receivedAtEpochMs: Int64
    let sample: ImuSample
}
